[RHELC-1089] Add result OVERRIDABLE where ask_to_continue() was removed

pr-watson opened the pull request:
This PR adds the "OVERRIDABLE" result in areas where the ask_to_continue() function was previously removed. The change allows for users to skip these checks by setting an environment variable.

# Note 
This branch was checked out from another PR #857 in review, since it had some required changes. The changes to look for are in pre_ponr_changes/handle_packages pre_ponr_changes/subscription.py, system_checks/package_updates.py and convert2rhel/subscription.py and all subsequent unit test files.


Jira Issues: [RHELC-1089](https://issues.redhat.com/browse/RHELC-1089)

Checklist

- [ ] PR has been tested manually in a VM (either author or reviewer)
- [ ] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title <!-- For a proper sync with Jira -->
- [x] GitHub label has been added to help with Release notes <!-- enhancement, bug-fix, no-changelog, security-hardening, breaking-change -->
- [x] PR title explains the change from the user's point of view
- [x] Code and tests are documented properly
- [x] The commits are squashed to as few commits as possible (without losing data) <!-- The commits can be squashed to 1 commit, but then we might lose data regarding moving something to a new file and then refactoring for example. Hence squash without losing data -->
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

pr-watson commented:
This branch was checked out from another PR https://github.com/oamg/convert2rhel/pull/841 since it had some required changes. The changes to look for are in pre_ponr_changes/handle_packages  pre_ponr_changes/subscription.py, system_checks/package_updates.py and convert2rhel/subscription.py and all subsequent unit test files.

abadger commented on convert2rhel/actions/pre_ponr_changes/subscription.py:
```diff
@@ -108,17 +108,6 @@ def run(self):
-                    message="%s repositories are not available - some packages"
-                    " may not be replaced and thus not supported." % ", ".join(non_available_repos),
-                )
-
```
Is this the check that Michal said we could remove now?

pr-watson replied:
Yes, it is apart of this ticket https://issues.redhat.com/browse/RHELC-1017

r0x0d commented on convert2rhel/unit_tests/__init__.py:
```diff
@@ -448,12 +448,17 @@ def assert_actions_result(instance, level=_NO_USER_VALUE, id=_NO_USER_VALUE, mes
     """Helper function to assert result set by Actions Framework."""
 
     if level != _NO_USER_VALUE:
+        print(instance.result.level)
+        print(STATUS_CODE[level])
         assert instance.result.level == STATUS_CODE[level]
 
     if id != _NO_USER_VALUE:
         assert instance.result.id == id
 
     if message != _NO_USER_VALUE:
+        print(instance.result.message)
+        print("\n")
+        print(message)
```
Don't forget to remove the debug statements later! 

```suggestion
        assert instance.result.level == STATUS_CODE[level]

    if id != _NO_USER_VALUE:
        assert instance.result.id == id

    if message != _NO_USER_VALUE:
```

r0x0d commented on convert2rhel/unit_tests/actions/pre_ponr_changes/kernel_modules_test.py:
```diff
@@ -181,6 +181,12 @@ def test_ensure_compatibility_of_kmods_check_env(
         " We will continue the conversion with the following kernel modules unavailable in RHEL:.*"
     )
     assert re.match(pattern=should_be_in_logs, string=caplog.records[-1].message, flags=re.MULTILINE | re.DOTALL)
+    # cannot assert exact action message contents as the kmods arrangement in the message is not static
+    assert "level=WARNING" in str(ensure_kernel_modules_compatibility_instance.messages)
+    assert "id=ALLOW_UNAVAILABLE_KERNEL_MODULES" in str(ensure_kernel_modules_compatibility_instance.messages)
+    assert "Detected 'CONVERT2RHEL_ALLOW_UNAVAILABLE_KMODS' environment variable." in str(
+        ensure_kernel_modules_compatibility_instance.messages
+    )
```
You could do it also like this: 

```python
    message = ensure_kernel_modules_compatibility_instance.messages[0]
    assert actions.STATUS_CODE["WARNING"] == message.level
    assert "ALLOW_UNAVAILABLE_KERNEL_MODULES" == message.id
    assert "Detected 'CONVERT2RHEL_ALLOW_UNAVAILABLE_KMODS' environment variable." in message.message
```

Which, in my opinion, is way more readable as well

r0x0d replied:
Since in this case you have a list of results, I would say that you don't necessarily need to on `assert_actions_result` for that assertation, and rather, you can do it manually like the above suggestion.

r0x0d replied:
Or you could also try to use the same strategy of your other checks using `set`

```python
    expected = set(
        (
            actions.ActionMessage(
                level="INFO",
                id="PACKAGE_UPDATES_CHECK_SKIP_NO_PUBLIC_REPOSITORIES",
                message=(
                    "Skipping the check because there are no publicly available Oracle Linux Server 8.6 repositories available."
                ),
            ),
        )
    )

    assert expected.issuperset(package_updates_action.messages)
    assert expected.issubset(package_updates_action.messages)
```

pr-watson replied:
I would do it with the subsets but the message inside of the ActionMessage is not static so it won't pass most of the time

r0x0d commented on convert2rhel/unit_tests/actions/system_checks/package_updates_test.py:
```diff
@@ -65,22 +79,150 @@ def test_check_package_updates(pretend_os, packages, exception, expected, monkey
+    monkeypatch.setattr(package_updates, "get_total_packages_to_update", value=get_total_packages_to_update_mock)
+    monkeypatch.setattr(package_updates, "get_total_packages_to_update", value=get_total_packages_to_update_mock)
+    package_updates_action.run()
+    print("start")
+    print(package_updates_action.result)
+    print("end")
```
Don't forget to remove those debug statements here too! 


```suggestion
    print(package_updates_action.result)
```

r0x0d commented on convert2rhel/unit_tests/actions/system_checks/rhel_compatible_kernel_test.py:
```diff
@@ -146,6 +239,48 @@ def test_bad_kernel_substring(kernel_release, exp_return):
+    ),
+)
+@centos8
+def test_bad_kernel_package_signature_success(
```
Why this was renamed?

r0x0d replied:
Also, it seems that `get_installed_pkg_objects` is not used in the test, maybe it is worth it to remove?

pr-watson replied:
get_installed_pkg_objects is used to mock information used in the check, without it the test fails

r0x0d reviewed (commented):
Another round of reviews and a few comments!

pr-watson commented on convert2rhel/backup.py:
```diff
@@ -434,10 +437,17 @@ def remove_pkgs(
+
+    if pkgs_failed_to_remove:
+        if critical:
+            loggerinst.critical("Error: Couldn't remove %s." % ", ".join(pkgs_failed_to_remove))
```
In the action that calls this function - handle_packages, we have a try/except block that will catch the logger critical and set an error result.
try:
            pkgs_to_remove = sorted(pkghandler._get_packages_to_remove(system_info.excluded_pkgs))
            pkgs_removed = sorted(pkghandler.remove_pkgs_unless_from_redhat(pkgs_to_remove))

            # TODO: Handling SystemExit here as way to speedup exception
            # handling and not refactor contents of the underlying function.
        except SystemExit as e:
            # TODO(r0x0d): Places where we raise SystemExit and need to be
            # changed to something more specific.
            #   - When we can't remove a package.
            self.set_result(level="ERROR", id="EXCLUDED_PACKAGE_REMOVAL_FAILED", message=str(e))
            return

abadger replied:
@pr-watson I think Rodolfo is saying that we can make this code raise a more specific exception than `SystemExit`.  Then in handle_packages, we can add another `except NewException:` block which does the proper `set_result` with a more accurate id and message.

pr-watson replied:
After discussion with Toshio we have agreed to leave this function as is since the logger critical was implemented before the action framework. For the specific exception being raised, that will be taken care of in a separate PR

pr-watson commented on convert2rhel/actions/system_checks/rhel_compatible_kernel.py:
```diff
@@ -43,57 +58,66 @@ def run(self):
+                check_function(system_info.booted_kernel)
+            except KernelIncompatibleError as e:
+                bad_kernel_message = str(e)
+                logger.warning(bad_kernel_message)
```
I suppose it doesn't add much value, I will remove this

pr-watson commented on convert2rhel/actions/system_checks/rhel_compatible_kernel.py:
```diff
@@ -43,57 +58,66 @@ def run(self):
+            ),
+        )
+
+    logger.debug(
+        "Booted kernel version '%s' corresponds to the version available in RHEL %d"
+        % (kernel_version, system_info.version.major)
+    )
```
I can add this

pr-watson commented on convert2rhel/unit_tests/actions/system_checks/dbus_test.py:
```diff
@@ -65,3 +65,22 @@ def test_check_dbus_is_running_not_running(monkeypatch, global_tool_opts, global
+
+
+def test_check_dbus_is_running_warning_message(
+    monkeypatch, global_tool_opts, global_system_info, dbus_is_running_action
```
Good catch, I will remove this

r0x0d commented on convert2rhel/actions/pre_ponr_changes/handle_packages.py:
```diff
@@ -60,16 +83,32 @@ def run(self):
         try:
-            pkgs_to_remove = pkghandler._get_packages_to_remove(system_info.excluded_pkgs)
-            pkghandler.remove_pkgs_unless_from_redhat(pkgs_to_remove)
+            pkgs_to_remove = sorted(pkghandler._get_packages_to_remove(system_info.excluded_pkgs))
```
@pr-watson

pr-watson replied:
Will do

r0x0d commented on convert2rhel/actions/system_checks/package_updates.py:
```diff
@@ -39,12 +40,25 @@ def run(self):
                 "Skipping the check because there are no publicly available %s %d.%d repositories available."
                 % (system_info.name, system_info.version.major, system_info.version.minor)
             )
+            self.add_message(
+                level="INFO",
+                id="PACKAGE_UPDATES_CHECK_SKIP_NO_PUBLIC_REPOSITORIES",
+                message=(
+                    "Skipping the check because there are no publicly available %s %d.%d repositories available."
+                    % (system_info.name, system_info.version.major, system_info.version.minor)
+                ),
+            )
```
Please, extract the raw message here and place it into a `message` variable, so it can be used in both the add_message and the logger call above.

r0x0d commented on convert2rhel/actions/system_checks/package_updates.py:
```diff
@@ -39,12 +40,25 @@ def run(self):
 
         if reposdir and not system_info.has_internet_access:
             logger.warning("Skipping the check as no internet connection has been detected.")
+            self.add_message(
+                level="WARNING",
+                id="PACKAGE_UPDATES_CHECK_SKIP_NO_INTERNET",
+                message="Skipping the check as no internet connection has been detected.",
+            )
```
Please, extract the raw message here and place it into a `message` variable, so it can be used in both the add_message and the logger call above.

r0x0d reviewed (commented):
Looks good to me. Only a couple of comments.

bocekm commented on convert2rhel/actions/system_checks/dbus.py:
```diff
@@ -35,6 +35,11 @@ def run(self):
         if tool_opts.no_rhsm:
             logger.info("Skipping the check because we have been asked not to subscribe this system to RHSM.")
+            self.add_message(
+                level="WARNING",
```
```suggestion
                level="INFO",
```
It's kind of an internal thing - convert2rhel needs dbus to work with RHSM. So this should be just info level because the customer doesn't care whether we're checking dbus when they are not using RHSM for the conversion.

bocekm replied:
I've taken my comment to the PR where this change has been introduced: https://github.com/oamg/convert2rhel/pull/878/files#r1300442040/

This PR just cherry-picks this change from that PR.

SpyTec commented:
Note from @danmyway, where it shows that commas are added everywhere

```
[2023-08-15T17:10:30+0000] TASK - [Convert: List third-party packages] *******************************

WARNING - Only packages signed by CentOS Linux are to be replaced. Red Hat support won't be provided for the following third party packages:

P, a, c, k, a, g, e,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  , V, e, n, d, o, r, /, P, a, c, k, a, g, e, r,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  , R, e, p, o, s, i, t, o, r, y, 

```

SpyTec commented:
Needs a bit of a rebase @pr-watson. Maybe we split this PR into a few more so we can merge each file separately (after merging the initial [convert2rhel/actions/__init__.py](https://github.com/oamg/convert2rhel/pull/878/files#diff-f2f9ed8f43c78514019a940b438afdcf05baca564c83c023829a3fbab0dc4798)

pr-watson commented:
> Needs a bit of a rebase @pr-watson. Maybe we split this PR into a few more so we can merge each file separately (after merging the initial [convert2rhel/actions/**init**.py](https://github.com/oamg/convert2rhel/pull/878/files#diff-f2f9ed8f43c78514019a940b438afdcf05baca564c83c023829a3fbab0dc4798)

Sure let me rebase and split this PR

pr-watson commented on convert2rhel/actions/system_checks/convert2rhel_latest.py:
```diff
@@ -60,6 +60,11 @@ def run(self):
 
         if not system_info.has_internet_access:
             logger.warning("Skipping the check because no internet connection has been detected.")
+            self.add_message(
+                level="WARNING",
+                id="CONVERT2RHEL_LATEST_CHECK_SKIP_NO_INTERNET",
+                message="Skipping the check because no internet connection has been detected.",
+            )
```
Noted, I'll update this

bocekm commented:
@pr-watson, rebase the PR, please.

pr-watson commented:
The required changes from this PR have been split into these two PRs - https://github.com/oamg/convert2rhel/pull/978 https://github.com/oamg/convert2rhel/pull/977. The rest of the changes have already been merged through https://github.com/oamg/convert2rhel/pull/888 so this PR can be closed
